README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# timevarcorr

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/timevarcorr)](https://CRAN.R-project.org/package=timevarcorr)
[![R-CMD-check](https://github.com/courtiol/timevarcorr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/courtiol/timevarcorr/actions/workflows/R-CMD-check.yaml)
[![test-coverage](https://github.com/courtiol/timevarcorr/actions/workflows/test-coverage.yaml/badge.svg)](https://github.com/courtiol/timevarcorr/actions/workflows/test-coverage.yaml)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

This R package aims at measuring how the correlation between 2
time-series changes over time, following the method described in Choi &
Shin (2021):

Choi, JE., Shin, D.W. Nonparametric estimation of time varying
correlation coefficient. J. Korean Stat. Soc. 50, 333–353 (2021).
<doi:10.1007/s42952-020-00073-6>

The chief idea is to perform a non-parametric kernel smoothing (using a
common bandwidth) of all underlying components required for the
computation of a correlation coefficient (i.e., $x$, $y$, $x^2$, $y^2$,
$x*y$).

The automatic selection procedure for the bandwidth parameter proposed
in the paper is implemented in this package. The same goes for the
computation of confidence intervals.

We also implemented the possibility to use Epanechnikov, Gaussian, or
box kernels, as well as to estimate either the Pearson or the Spearman
correlation coefficient.

## Installation

You can install the CRAN version of **timevarcorr** with:

``` r
install.packages("timevarcorr")
```

You can install the development version of **timevarcorr** from GitHub
with:

``` r
# install.packages("remotes") ## uncomment and run if you don't have this package installed
remotes::install_github("courtiol/timevarcorr")
```

That should suffice!

Note that this package relies so far on only one direct dependency –
[**lpridge**](https://github.com/cran/lpridge) – which itself depends on
nothing but a plain R install.

Nonetheless, in some of the examples below, we also rely on
[**dplyr**](https://dplyr.tidyverse.org) and
[**ggplot2**](https://ggplot2.tidyverse.org), so you would need to
install these packages as well to reproduce the content of this README:

``` r
install.packages("dplyr")
install.packages("ggplot2")
```

## Examples

The main function of this package is called `tcor` and its documentation
is available here:

``` r
help(tcor, package = timevarcorr)
```

Here is a very simple example using base-R syntax:

``` r
library(timevarcorr)
#> timevarcorr loaded; type ?tcor for help on this package.

d <- stockprice[1:500, ]
example1 <- with(d, tcor(x = SP500, y = FTSE100, t = DateID, kernel = "normal"))
#> 
#> You may use several CPU cores for faster computation by calling `options('mc.cores' = XX)` with `XX` corresponding to the number of CPU cores to be used.
#> h selected using LOO-CV = 60.9
#> Bandwidth automatic selection completed in 9.1 seconds
plot(example1, type = "l")
```

<img src="man/figures/README-example-1.png" width="70%" style="display: block; margin: auto;" />

Here is the same example using tidyverse syntax (with confidence
interval):

``` r
library(dplyr)
library(ggplot2)

d |> 
  reframe(tcor(x = SP500, y = FTSE100, t = DateID,
               kernel = "normal", CI = TRUE)) |>
  ggplot() +
    aes(x = t, y = r, ymin = lwr, ymax = upr) +
    geom_ribbon(fill = "grey") +
    geom_line() +
    labs(title = "SP500 vs FTSE100", x = "Time", y = "Correlation") +
    theme_classic()
```

<img src="man/figures/README-example2-1.png" width="70%" style="display: block; margin: auto;" />

And now, the same example showing gaps of observations in the time
series:

``` r
d |> 
  reframe(tcor(x = SP500, y = FTSE100, t = DateID,
               kernel = "normal", CI = TRUE, keep.missing = TRUE)) |>
  ggplot() +
    aes(x = t, y = r, ymin = lwr, ymax = upr) +
    geom_ribbon(fill = "grey") +
    geom_line() +
    labs(title = "SP500 vs FTSE100", x = "Time", y = "Correlation") +
    theme_classic()
#> 
#> You may use several CPU cores for faster computation by calling `options('mc.cores' = XX)` with `XX` corresponding to the number of CPU cores to be used.
#> h selected using LOO-CV = 60.9
#> Bandwidth automatic selection completed in 8.7 seconds
```

<img src="man/figures/README-example3-1.png" width="70%" style="display: block; margin: auto;" />

You can do more. For example, you can use other kernels, fix the
bandwidth manually, or use the Spearman’s rather than the Pearson’s
correlation coefficient:

``` r
example2 <- with(d, tcor(x = SP500, y = FTSE100, t = DateID,
                 cor.method = "spearman", kernel = "box", h = 10))
plot(example2, type = "l")
```

<img src="man/figures/README-example4-1.png" width="70%" style="display: block; margin: auto;" />

You can also test the difference in correlation coefficients between two
time points:

``` r
example3 <- with(d, tcor(x = SP500, y = FTSE100, t = DateID, kernel = "normal", CI = TRUE))
#> 
#> You may use several CPU cores for faster computation by calling `options('mc.cores' = XX)` with `XX` corresponding to the number of CPU cores to be used.
#> h selected using LOO-CV = 60.9
#> Bandwidth automatic selection completed in 8.9 seconds
test_equality(example3, t1 = "2000-05-02", t2 = "2001-05-02")
#>           t1        r1         t2     r2   delta_r SE_delta_r   T_stat  df
#> 1 2000-05-02 0.4354492 2001-05-02 0.5722 0.1367509  0.1224746 1.116565 910
#>          p
#> 1 0.264475
```

Or you can test if specific time points (or all) differ from a reference
value:

``` r
test_ref(example3, t = c("2000-05-02", "2001-05-02"), r_ref = 0.5)
#>            t         r r_ref     delta_r SE_delta_r     T_stat  df         p
#> 1 2000-05-02 0.4354492   0.5 -0.06455083 0.10082705 -0.6402134 910 0.5221950
#> 2 2001-05-02 0.5722000   0.5  0.07220003 0.06952644  1.0384542 910 0.2993345
#>   p_adjustment
#> 1         none
#> 2         none
```

## Devel corner

This README file has been compiled using `devtools::build_readme()`,
with the following setup:

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.3.1 (2023-06-16)
#>  os       Ubuntu 22.04.3 LTS
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Europe/Berlin
#>  date     2023-11-06
#>  pandoc   3.1.8 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/x86_64/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  cachem        1.0.8   2023-05-01 [4] RSPM (R 4.2.0)
#>  callr         3.7.3   2022-11-02 [4] RSPM (R 4.2.0)
#>  cli           3.6.1   2023-03-23 [4] RSPM (R 4.2.0)
#>  colorspace    2.1-0   2023-01-23 [4] RSPM (R 4.2.0)
#>  crayon        1.5.2   2022-09-29 [4] RSPM (R 4.2.0)
#>  devtools      2.4.5   2022-10-11 [4] RSPM (R 4.2.0)
#>  digest        0.6.33  2023-07-07 [4] RSPM (R 4.2.0)
#>  dplyr       * 1.1.3   2023-09-03 [4] RSPM (R 4.3.0)
#>  ellipsis      0.3.2   2021-04-29 [4] RSPM (R 4.2.0)
#>  evaluate      0.23    2023-11-01 [4] RSPM (R 4.3.0)
#>  fansi         1.0.5   2023-10-08 [4] RSPM (R 4.3.0)
#>  farver        2.1.1   2022-07-06 [4] RSPM (R 4.2.0)
#>  fastmap       1.1.1   2023-02-24 [4] RSPM (R 4.2.0)
#>  fs            1.6.3   2023-07-20 [4] RSPM (R 4.2.0)
#>  generics      0.1.3   2022-07-05 [4] RSPM (R 4.2.0)
#>  ggplot2     * 3.4.4   2023-10-12 [4] RSPM (R 4.3.0)
#>  glue          1.6.2   2022-02-24 [4] RSPM (R 4.2.0)
#>  gtable        0.3.4   2023-08-21 [4] RSPM (R 4.2.0)
#>  highr         0.10    2022-12-22 [4] RSPM (R 4.2.0)
#>  htmltools     0.5.6.1 2023-10-06 [4] RSPM (R 4.3.0)
#>  htmlwidgets   1.6.2   2023-03-17 [4] RSPM (R 4.2.0)
#>  httpuv        1.6.12  2023-10-23 [4] RSPM (R 4.3.0)
#>  knitr         1.45    2023-10-30 [4] RSPM (R 4.3.0)
#>  labeling      0.4.3   2023-08-29 [4] RSPM (R 4.2.0)
#>  later         1.3.1   2023-05-02 [4] RSPM (R 4.2.0)
#>  lifecycle     1.0.3   2022-10-07 [4] RSPM (R 4.2.0)
#>  magrittr      2.0.3   2022-03-30 [4] RSPM (R 4.2.0)
#>  memoise       2.0.1   2021-11-26 [4] RSPM (R 4.2.0)
#>  mime          0.12    2021-09-28 [4] RSPM (R 4.2.0)
#>  miniUI        0.1.1.1 2018-05-18 [4] RSPM (R 4.2.0)
#>  munsell       0.5.0   2018-06-12 [4] RSPM (R 4.2.0)
#>  pillar        1.9.0   2023-03-22 [4] RSPM (R 4.2.0)
#>  pkgbuild      1.4.2   2023-06-26 [4] RSPM (R 4.2.0)
#>  pkgconfig     2.0.3   2019-09-22 [4] RSPM (R 4.2.0)
#>  pkgload       1.3.3   2023-09-22 [4] RSPM (R 4.3.0)
#>  prettyunits   1.2.0   2023-09-24 [4] RSPM (R 4.3.0)
#>  processx      3.8.2   2023-06-30 [4] RSPM (R 4.2.0)
#>  profvis       0.3.8   2023-05-02 [4] RSPM (R 4.2.0)
#>  promises      1.2.1   2023-08-10 [4] RSPM (R 4.2.0)
#>  ps            1.7.5   2023-04-18 [4] RSPM (R 4.3.0)
#>  purrr         1.0.2   2023-08-10 [4] RSPM (R 4.2.0)
#>  R6            2.5.1   2021-08-19 [4] RSPM (R 4.2.0)
#>  Rcpp          1.0.11  2023-07-06 [4] RSPM (R 4.2.0)
#>  remotes       2.4.2.1 2023-07-18 [4] RSPM (R 4.2.0)
#>  rlang         1.1.1   2023-04-28 [4] RSPM (R 4.2.0)
#>  rmarkdown     2.25    2023-09-18 [4] RSPM (R 4.3.0)
#>  rstudioapi    0.15.0  2023-07-07 [4] RSPM (R 4.2.0)
#>  scales        1.2.1   2022-08-20 [4] RSPM (R 4.2.0)
#>  sessioninfo   1.2.2   2021-12-06 [4] RSPM (R 4.2.0)
#>  shiny         1.7.5.1 2023-10-14 [4] RSPM (R 4.3.0)
#>  stringi       1.7.12  2023-01-11 [4] RSPM (R 4.2.0)
#>  stringr       1.5.0   2022-12-02 [4] RSPM (R 4.2.0)
#>  tibble        3.2.1   2023-03-20 [4] RSPM (R 4.3.0)
#>  tidyselect    1.2.0   2022-10-10 [4] RSPM (R 4.2.0)
#>  timevarcorr * 0.1.1   2023-11-06 [1] local
#>  urlchecker    1.0.1   2021-11-30 [4] RSPM (R 4.2.0)
#>  usethis       2.2.2   2023-07-06 [4] RSPM (R 4.2.0)
#>  utf8          1.2.4   2023-10-22 [4] RSPM (R 4.3.0)
#>  vctrs         0.6.4   2023-10-12 [4] RSPM (R 4.3.0)
#>  withr         2.5.2   2023-10-30 [4] RSPM (R 4.3.0)
#>  xfun          0.41    2023-11-01 [4] RSPM (R 4.3.0)
#>  xtable        1.8-4   2019-04-21 [4] RSPM (R 4.2.0)
#>  yaml          2.3.7   2023-01-23 [4] RSPM (R 4.2.0)
#> 
#>  [1] /tmp/RtmpZf1Gw1/temp_libpath1049fd45c346be
#>  [2] /home/courtiol/R/x86_64-pc-linux-gnu-library/4.3
#>  [3] /usr/local/lib/R/site-library
#>  [4] /usr/lib/R/site-library
#>  [5] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```


NEWS.md
# timevarcorr (development version)

# timevarcorr 0.1.1 (First CRAN release)

New submission to CRAN after revisions (accepted).

* Add Choi & Shin (2021) in DESCRIPTION.
* Improve documentation (returns more detailed and use of LaTeX).

# timevarcorr 0.1.0

Initial CRAN submission (rejected).

# timevarcorr 0.0.0.9006 and earlier

Prepare package for CRAN release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(timevarcorr)

test_check("timevarcorr")


tests/testthat/test-calc_rho.R
test_that("dates are not lost", {
  job <- with(stockprice[1:6, ], calc_rho(x = SP500, y = FTSE100, t = DateID, h = 20))
  expect_true(inherits(job$t, "Date"))
})

test_that("order of time points do not alter the result", {
  ordered <- with(stockprice[1:6, ], calc_rho(x = SP500, y = FTSE100, t = DateID, h = 20))
  disorderd <- with(stockprice[6:1, ], calc_rho(x = SP500, y = FTSE100, t = DateID, h = 20))
  reordered <- disorderd[6:1, ]
  rownames(reordered) <- NULL
  expect_equal(ordered, reordered)
})

test_that("correlations under `h = Inf` match time-invariant correlations", {
  rho_inf <- unique(with(stockprice[1:6, ], calc_rho(x = SP500, y = FTSE100, t = DateID, h = Inf, kernel = "box"))$rho)
  rho_inv <- cor(stockprice[1:6, "SP500"], stockprice[1:6, "FTSE100"])[[1]]
  expect_equal(rho_inf, rho_inv)
})


R/CI.R
#' Internal functions for the computation of confidence intervals
#'
#' These functions compute the different terms required for [`tcor()`] to compute the confidence
#' interval around the time-varying correlation coefficient. These terms are defined in Choi & Shin (2021).
#'
#' @seealso [`tcor()`]
#' @name CI
#'
#' @references
#' Choi, JE., Shin, D.W. Nonparametric estimation of time varying correlation coefficient.
#' J. Korean Stat. Soc. 50, 333–353 (2021). \doi{10.1007/s42952-020-00073-6}
#'
#' Andrews, D. W. K. Heteroskedasticity and autocorrelation consistent covariance matrix estimation.
#' Econometrica: Journal of the Econometric Society, 817-858 (1991).
#'
#' @return
#' - `calc_H()` returns a 5 x 5 x \eqn{t} array of elements of class numeric, which corresponds to \eqn{\hat{H_t}} in Choi & Shin (2021).
#' - `calc_e()` returns a \eqn{t} x 5 matrix of elements of class numeric storing the residuals, which corresponds to \eqn{\hat{e}_t} in Choi & Shin (2021).
#' - `calc_Gamma()` returns a 5 x 5 matrix of elements of class numeric, which corresponds to \eqn{\hat{\Gamma}_l} in Choi & Shin (2021).
#' - `calc_GammaINF()` returns a 5 x 5 matrix of elements of class numeric, which corresponds to \eqn{\hat{\Gamma}^\infty} in Choi & Shin (2021).
#' - `calc_L_And()` returns a scalar of class numeric, which corresponds to \eqn{L_{And}} in Choi & Shin (2021).
#' - `calc_D()` returns a \eqn{t} x 5 matrix of elements of class numeric storing the residuals, which corresponds to \eqn{D_t} in Choi & Shin (2021).
#' - `calc_SE()` returns a vector of length \eqn{t} of elements of class numeric, which corresponds to \eqn{se(\hat{\rho}_t(h))} in Choi & Shin (2021).
#'
#' @examples
#' rho_obj <- with(na.omit(stockprice),
#'                 calc_rho(x = SP500, y = FTSE100, t = DateID, h = 20, kernel = "box"))
#' head(rho_obj)
#'
NULL


#' @describeIn CI computes the \eqn{\hat{H_t}} array.
#'
#' \eqn{\hat{H_t}} is a component needed to compute confidence intervals;
#' \eqn{H_t} is defined in eq. 6 from Choi & Shin (2021).
#'
#' @export
#' @param smoothed_obj an object created with [`calc_rho`].
#'
#' @examples
#' ## Computing \eqn{\hat{H_t}}
#'
#' H <- calc_H(smoothed_obj = rho_obj)
#' H[, , 1:2] # H array for the first two time points
#'
calc_H <- function(smoothed_obj) {
  res <- array(0, dim = c(5, 5, nrow(smoothed_obj)))

  res[1, 1, ] <- 2*smoothed_obj$x_smoothed*smoothed_obj$sd_x_smoothed
  res[1, 2, ] <- 2*smoothed_obj$y_smoothed*smoothed_obj$sd_y*smoothed_obj$rho_smoothed
  res[1, 3, ] <- smoothed_obj$sd_x_smoothed
  res[1, 4, ] <- smoothed_obj$sd_y_smoothed*smoothed_obj$rho_smoothed
  res[1, 5, ] <- smoothed_obj$x_smoothed*smoothed_obj$sd_y_smoothed*smoothed_obj$rho_smoothed + smoothed_obj$y_smoothed*smoothed_obj$sd_x_smoothed

  res[2, 2, ] <- 2*smoothed_obj$y_smoothed*smoothed_obj$sd_y_smoothed*sqrt(1 - smoothed_obj$rho_smoothed^2)
  res[2, 4, ] <- smoothed_obj$sd_y_smoothed*sqrt(1 - smoothed_obj$rho_smoothed^2)
  res[2, 5, ] <- smoothed_obj$x_smoothed*smoothed_obj$sd_y_smoothed*sqrt(1 - smoothed_obj$rho_smoothed^2)

  res[3, 1, ] <- smoothed_obj$sd_x_smoothed^2
  res[3, 2, ] <- smoothed_obj$sd_y_smoothed^2*smoothed_obj$rho_smoothed^2
  res[3, 5, ] <- smoothed_obj$sd_x_smoothed*smoothed_obj$sd_y_smoothed*smoothed_obj$rho_smoothed

  res[4, 2, ] <- smoothed_obj$sd_y_smoothed^2*(smoothed_obj$rho_smoothed^2 - 1)

  res[5, 2, ] <- 2*smoothed_obj$sd_y_smoothed^2*smoothed_obj$rho_smoothed*sqrt(1 - smoothed_obj$rho_smoothed^2)
  res[5, 5, ] <- sqrt(1 - smoothed_obj$rho_smoothed^2)*smoothed_obj$sd_x_smoothed*smoothed_obj$sd_y_smoothed

  res
}


#' @describeIn CI computes \eqn{\hat{e}_t}.
#'
#' \eqn{\hat{e}_t} is defined in eq. 9 from Choi & Shin (2021).
#'
#' @export
#' @param H an object created with `calc_H`.
#'
#' @examples
#' ## Computing \eqn{\hat{e}_t}
#'
#' e <- calc_e(smoothed_obj = rho_obj, H = H)
#' head(e) # e matrix for the first six time points
#'
calc_e <- function(smoothed_obj, H) {
  res <- matrix(0, ncol = 5, nrow = dim(H)[3])
  for (i in seq_len(dim(H)[3])) {
    Ut   <- t(smoothed_obj[i, c("x2", "y2", "x", "y", "xy"), drop = FALSE])
    mu_t <- t(smoothed_obj[i, c("x2_smoothed", "y2_smoothed", "x_smoothed", "y_smoothed", "xy_smoothed"), drop = FALSE])
    res[i, ] <- tryCatch({solve(t(H[, , i])) %*% (Ut - mu_t)}, error = function(e) stop("The bandwidth parameter `h` is too small. CI cannot be computed."))
  }
  colnames(res) <- paste0(c("x2", "y2", "x", "y", "xy"), "_resid")
  res
}


#' @describeIn CI computes \eqn{\hat{\Gamma}_l}.
#'
#' \eqn{\hat{\Gamma}_l} is defined in eq. 9 from Choi & Shin (2021).
#'
#' @export
#' @param e an object created with `calc_e`.
#' @param l a scalar indicating a number of time points.
#'
#' @examples
#' ## Computing \eqn{\hat{\Gamma}_l}
#'
#' calc_Gamma(e = e, l = 3)
#'
calc_Gamma <- function(e, l) {
  res <- matrix(0, nrow = 5, ncol = 5)
  for (t in seq_len(nrow(e) - l)) {
    res <- res + t(e[t, , drop = FALSE]) %*% e[t + l, , drop = FALSE]
  }
  res/nrow(e)
}


#' @describeIn CI computes \eqn{\hat{\Gamma}^\infty}.
#'
#' \eqn{\hat{\Gamma}^\infty} is the long run variance estimator, defined in eq. 9 from Choi & Shin (2021).
#'
#' @export
#' @param L a scalar indicating a bandwidth parameter.
#'
#' @examples
#' ## Computing \eqn{\hat{\Gamma}^\infty}
#'
#' calc_GammaINF(e = e, L = 2)
#'
calc_GammaINF <- function(e, L) {

  Gamma_0 <- calc_Gamma(e = e, l = 0)

  sum_term <- matrix(0, nrow = 5, ncol = 5)
  for (l in seq_len(L)) {
    Gamma_l <- calc_Gamma(e = e, l = l)
    sum_term <- sum_term + (1 - l/L) * (Gamma_l + t(Gamma_l))
  }

  Gamma_0 + sum_term
}


#' @describeIn CI computes \eqn{L_{And}}.
#'
#' \eqn{L_{And}} is defined in Choi & Shin (2021, p 342).
#' It also corresponds to \eqn{S_T^*}, eq 5.3 in Andrews (1991).
#'
#' @export
#' @param AR.method character string specifying the method to fit the autoregressive model used to compute \eqn{\hat{\gamma}_1} in \eqn{L_{And}} (see [`stats::ar`] for details).
#'
#' @examples
#' ## Computing \eqn{L_{And}}
#'
#' calc_L_And(e = e)
#' sapply(c("yule-walker", "burg", "ols", "mle", "yw"),
#'        function(m) calc_L_And(e = e, AR.method = m)) ## comparing AR.methods
#'
calc_L_And <- function(e, AR.method = c("yule-walker", "burg", "ols", "mle", "yw")) {
  AR.method <- match.arg(AR.method)
  gamma_1 <- stats::ar(rowSums(e), method = AR.method, order.max = 1L)$ar
  if (length(gamma_1) == 0) return(0)
  1.1447*((4*nrow(e)*gamma_1^2)/((1 - gamma_1^2)^2))^(1/3)
}


#' @describeIn CI computes \eqn{D_t}.
#'
#' \eqn{D_t} is defined in Choi & Shin (2021, p 338).
#'
#' @export
#'
#' @examples
#' ## Computing \eqn{D_t}
#'
#' D <- calc_D(smoothed_obj = rho_obj)
#' head(D) # D matrix for the first six time points
#'
calc_D <- function(smoothed_obj) {
  res <- matrix(0, nrow = nrow(smoothed_obj), ncol = 5)
  res[, 1] <- -1/2*(smoothed_obj$xy_smoothed - smoothed_obj$x_smoothed*smoothed_obj$y_smoothed)/(smoothed_obj$sd_x_smoothed^3*smoothed_obj$sd_y_smoothed)
  res[, 2] <- -1/2*(smoothed_obj$xy_smoothed - smoothed_obj$x_smoothed*smoothed_obj$y_smoothed)/(smoothed_obj$sd_x_smoothed*smoothed_obj$sd_y_smoothed^3)
  res[, 3] <- (-smoothed_obj$y_smoothed*smoothed_obj$sd_x_smoothed^2 + smoothed_obj$x_smoothed*(smoothed_obj$xy_smoothed - smoothed_obj$x_smoothed*smoothed_obj$y_smoothed))/(smoothed_obj$sd_x_smoothed^3*smoothed_obj$sd_y_smoothed)
  res[, 4] <- (-smoothed_obj$x_smoothed*smoothed_obj$sd_y_smoothed^2 + smoothed_obj$y_smoothed*(smoothed_obj$xy_smoothed - smoothed_obj$x_smoothed*smoothed_obj$y_smoothed))/(smoothed_obj$sd_x_smoothed*smoothed_obj$sd_y_smoothed^3)
  res[, 5] <- 1/(smoothed_obj$sd_x_smoothed*smoothed_obj$sd_y_smoothed)
  res
}


#' @describeIn CI computes \eqn{se(\hat{\rho}_t(h))}.
#'
#' The standard deviation of the time-varying correlation (\eqn{se(\hat{\rho}_t(h))}) is defined in eq. 8 from Choi & Shin (2021).
#' It depends on \eqn{D_{Lt}}, \eqn{D_{Mt}} & \eqn{D_{Ut}}, themselves defined in Choi & Shin (2021, p 337 & 339).
#' The \eqn{D_{Xt}} terms are all computed within the function since they all rely on the same components.
#'
#' @export
#' @inheritParams kern_smooth
#'
#' @examples
#' ## Computing \eqn{se(\hat{\rho}_t(h))}
#' # nb: takes a few seconds to run
#'
#' run <- FALSE ## change to TRUE to run the example
#' if (in_pkgdown() || run) {
#'
#' SE <- calc_SE(smoothed_obj = rho_obj, h = 50)
#' head(SE) # SE vector for the first six time points
#'
#' }
#'
#'
calc_SE <- function(smoothed_obj, h, AR.method = c("yule-walker", "burg", "ols", "mle", "yw")) {

  ## Bartlett kernel
  bk <- function(x) pmax(0, 1 - abs(x)) ## 1 - abs(x) if abs(x) < 1, 0 otherwise
  bk2 <- function(x) bk(x)^2

  ## Common terms
  H <- calc_H(smoothed_obj = smoothed_obj)
  e <- calc_e(smoothed_obj = smoothed_obj, H = H)
  D <- calc_D(smoothed_obj = smoothed_obj)
  L_And <- calc_L_And(e = e, AR.method = AR.method)
  GammaINF <- calc_GammaINF(e = e, L = L_And)
  N <- nrow(smoothed_obj)

  SE <- numeric(N)
  for (t in seq_len(N)) {
    Dt <- D[t, ]
    Ht <- H[, , t]

    ## boundary cases at the beginning of the time series
    if (t < h) {
      c <- t/h
      DLt <- stats::integrate(bk, -c, 1)$value^-2*t(Ht) %*% GammaINF %*% Ht * stats::integrate(bk2, -c, 1)$value
      SE[t] <- sqrt(t(Dt) %*% DLt %*% Dt/h)
    ## boundary cases at the end of the time series
    } else if (t > N - h) {
      c <- (N - t)/h
      DUt <- stats::integrate(bk, -1, c)$value^-2*t(Ht) %*% GammaINF %*% Ht * stats::integrate(bk2, -1, c)$value
      SE[t] <- sqrt(t(Dt) %*% DUt %*% Dt/h)
    ## non-boundary cases
    } else {
      DMt <- t(Ht) %*% GammaINF %*% Ht * stats::integrate(bk2, -1, 1)$value
      SE[t] <- sqrt(t(Dt) %*% DMt %*% Dt/h)
    }
  }
  SE
}


R/CV.R
#' @describeIn tcor Internal function selecting the optimal bandwidth parameter `h`.
#'
#' See **Bandwidth selection** for details.
#'
#' @order 3
#'
#'
#' @export
#'
select_h <- function(x, y, t = seq_along(x), cor.method = c("pearson", "spearman"),
                     kernel = c("epanechnikov", "box", "normal"), param_smoother = list(),
                     verbose = FALSE) {

  CV_error <- NA # default value for export if not computed

  ## check nb.cores input
  if (is.null(options("mc.cores")[[1]])) {
    nb.cores <- 1
  } else {
    nb.cores <- options("mc.cores")[[1]]
  }

  if (Sys.info()[['sysname']] != "Windows" && parallel::detectCores() < nb.cores) {
    stop(paste("\nYour computer does not allow such a large value for the argument `nb.cores`. The maximum value you may consider is", parallel::detectCores(), ".\n"))
  }

  if (Sys.info()[['sysname']] != "Windows" && parallel::detectCores() > 1 && nb.cores == 1) {
    message("\nYou may set `nb.cores` to a number higher than 1 for faster computation.\n")
  }

  if (Sys.info()[['sysname']] == "Windows" && nb.cores > 1) {
    message("\nThe argument `nb.cores` is ignore on Windows-based infrastructure.\n")
  }

  time <- system.time({

    h_selection <- "LOO-CV"

    if (length(x) > 500) message("Bandwidth selection using LOO-CV... (may take a while)")

    ## define function for performing leave-one-out cross-validation ## TODO: extract code into standalone function
    CV <- function(h) {
      CVi <- mclapply(seq_along(t), function(oob) {
        obj <- calc_rho(x = x[-oob], y = y[-oob], t = t[-oob], h = h, t.for.pred = t[oob],
                        cor.method = cor.method, kernel = kernel, param_smoother = param_smoother)
        (obj$rho_smoothed - ((x[oob] - obj$x_smoothed) * (y[oob] - obj$y_smoothed)) / (obj$sd_x_smoothed * obj$sd_y_smoothed))^2
      })
      CVi_num <- as.numeric(CVi)
      failing <- is.na(CVi_num)
      res <- mean(CVi_num[!failing])
      if (verbose) {
        print(paste("h =", round(h, digits = 1L), "    # failing =", sum(failing),"    CV =", res))
      }
      res
    }

    ## estimate h by cross-validation
    h_max <- 8*sqrt(length(x))
    opt <- stats::optimize(CV, interval = c(1, h_max), tol = 1) ## TODO: check if other optimizer would be best

    ## if h_max is best, use elbow criterion instead
    h <- opt$minimum
    CV_error <- opt$objective
    print(paste("h selected using LOO-CV =", round(h, digits = 1L)))
    CV_bound <- CV(h_max)

    if (CV_bound <= opt$objective) {
      CV_error <- NA # remove stored value since not used in this case
      h_selection <- "elbow criterion"

      if (length(x) > 500) message("Bandwidth selection using elbow criterion... (may take a while)")

      RMSE_Lh0 <- function(h0) {
        d <- data.frame(h = 1:h0)
        d$CV_h <- sapply(h, CV)
        fit_Lh0 <- stats::lm(CV_h ~ h, data = d)
        sqrt(mean(stats::residuals(fit_Lh0)^2))
      }

      RMSE_Rh0 <- function(h0) {
        d <- data.frame(h = (h0 + 1):h_max)
        d$CV_h <- sapply(h, CV)
        fit_Rh0 <- stats::lm(CV_h ~ h, data = d)
        sqrt(mean(stats::residuals(fit_Rh0)^2))
      }

      RMSE_h0 <- function(h0) {
        (h0 - 1)/(h_max - 1)*RMSE_Lh0(h0) + (h_max - h0)/(h_max - 1)*RMSE_Rh0(h0)
      }

      opt <- stats::optimize(RMSE_h0, interval = c(3, h_max - 1), tol = 1)
      h <- opt$minimum
      print(paste("h selected using elbow criterion =", round(h, digits = 1L)))
    }
  })
  list(h = h, h_selection = h_selection, CV_error = CV_error, time = time)
}


R/zzz.R
#' Display welcome message
#'
#' This function should not be called by the user.
#' It displays a message when the package is being loaded.
#'
#' @param libname argument needed but automatically defined.
#' @param pkgname argument needed but automatically defined.
#'
#' @export
#' @return nothing (invisible NULL).
#'
.onAttach <- function(libname, pkgname) {
  packageStartupMessage(## display message
    'timevarcorr loaded; type ?tcor for help on this package.',
    '\n'
  )
}


#' Determine if the package is being used by pkgdown
#'
#' This function should not be called by the user.
#' It allows to run some examples conditionally to being used by pkgdown.
#' Code copied from [pkgdown::in_pkgdown()].
#'
#' @export
#'
#' @return a logical value (`TRUE` or `FALSE`).
#'
in_pkgdown <- function() {
  identical(Sys.getenv("IN_PKGDOWN"), "true")
}


R/data.R
#' Daily Closing Prices of Major European Stock Indices, April 2000--December 2017
#'
#' A dataset containing the stockmarket returns between 2000-04-03 and 2017-12-05.
#' This dataset is very close to the one used by Choi & Shin (2021), although not
#' strictly identical. It has been produced by the Oxford-Man Institute of Quantitative Finance.
#'
#' @format A data frame with 4618 rows and 7 variables:
#' \describe{
#'   \item{DateID}{a vector of `Date`.}
#'   \item{SP500}{a numeric vector of the stockmarket return for the S&P 500 Index.}
#'   \item{FTSE100}{a numeric vector of the stockmarket return for the FTSE 100.}
#'   \item{Nikkei}{a numeric vector of the stockmarket return for the Nikkei 225.}
#'   \item{DAX}{a numeric vector of the stockmarket return for the German stock index.}
#'   \item{NASDAQ}{a numeric vector of the stockmarket return for the Nasdaq Stock Market.}
#'   \item{Event}{a character string of particular events that have impacted the stockmarket, as in Choi & Shin (2021).}
#' }
#'
#' @source
#' The file was downloaded from the "Oxford-Man Institute's realized library", which no longer exists.
#' At the time, the raw data file was named "oxfordmanrealizedvolatilityindices-0.2-final.zip".
#'
#' @seealso
#' [`datasets::EuStockMarkets`] for a similar dataset, albeit formatted differently.
#'
#' @references
#' Heber, Gerd, Asger Lunde, Neil Shephard and Kevin Sheppard (2009)
#' "Oxford-Man Institute's realized library", Oxford-Man Institute, University of Oxford.
#'
#' Choi, JE., Shin, D.W. Nonparametric estimation of time varying correlation coefficient.
#' J. Korean Stat. Soc. 50, 333–353 (2021). \doi{10.1007/s42952-020-00073-6}
#'
"stockprice"
